### PR TITLE
[Topic Recognition]: ESPCN for Super-Resolution

### PR DESCRIPTION
**Pull request by yabbadabadoo**
This PR introduces the Efficient Sub-Pixel Convolutional Neural Network (ESPCN) model for super-resolution tasks on the ADNI dataset. This model was trained and tested on Rangpur, training at 20 epochs took around 19 minutes, reconstructed image displays a sharper super-resolved image. 

Changes:
- Added the `ESPCN` model class with feature extraction and sub-pixel convolution layers.
- Integrated the model into our training loop, including loss functions and optimizers.
- Included PSNR computation for model evaluation.
- Adjusted hyperparameters for optimal performance based on initial tests.

Files:
- train.py: Contains training loop and saving of the trained ESPCN model
- dataset.py: Transformation and loading of the ADNI dataset (ensure that your dataset fits the structure in the README file)
- predict.py: Generates a side-by-side comparison of the downsampled image (PyTorch Resize), upsampled image and the original image. This is our test script to ascertain the quality of upsampling.
- modules.py: Contains the ESPCN model
- README.md: Documentation detailing this project.
- flow.png: image detailing the convolution process
- outputimg.png: output image from predict.py
- plots.png: training plots
- subpix.png: image of sub-pixels

Testing:
- The model has been trained for 20 epochs, showing promising results in terms of PSNR. Further details in the README file.
- Manual inspection of the super-resolved images shows significant improvement over our previous model.

Additional Notes:
- Potential future work includes fine-tuning the model on domain-specific datasets for even better performance.
- Files in this PR are located in: recognition/superres_46501635 (all the files committed are in this folder ONLY)

Feedback and suggestions for improvement are welcome and appreciated.

### CODE REVIEW

**Comment by uqmarlonbran** [comment]
**This is an initial inspection - no action is required yet**

**Difficulty:** Normal

**General Comments:** Nice approach overall. README seems to have detail regarding PixelCNN, contrast not quite the same as the target image.

**[Recognition Problem]**
"modules.py":

- PyTorch implementation of PixelCNN. All looks ok.
- Nice use of Sequential Module.

"dataset.py":

- Custom dataset to deal with AD and NC folder structure.
- No validation set, only train and test.

"train.py":

- Training loop looks ok, keeps track of training and "validation" losses; validation is performed on the test set.
- Provides some figures for training plots.
- Computes metrics on train and test sets.

"predict.py":

- Correctly loads model weights.
- Just shows a series of sample reconstructions.

"README.MD":

- Provides file structure.
- Provides brief overview of network architecture and approach.
- Details the importance/benefit of sub-pixel convolution.
- Provides dependencies.
- Provides some usage examples.
- Provides training loss curves. Somewhat erratic, maybe due to high starting learning rate.
- Output images look ok, a little less detail than other submissions. Contrast a little different.

**[Commit Log]**
Commit log looks ok.

**Comment by shakes76** [comment]
# Marking
## Good Practice (Design/Commenting, TF/Torch Usage)
Adequate design and implementation
Good spacing and comments
Header blocks missing -1
## Recognition Problem 
Solves problem 
Driver Script present
File structure present
Shows Usage & Demo & Visualisation & Data usage
Module present
Commenting
No Data leakage, no validation set -1
Difficulty: Normal -5
## Commit Log
Meaningful commit messages
Progressive commits used
## Documentation
ReadMe acceptable
Model/technical explanation
Good Description and Comments 
Markdown used and PDF submitted
## Pull Request
Successful Pull Request (Working Algorithm Delivered on Time in Correct Branch) 
No Feedback required
Request Description good
